ansible_rulebook/terminal.py

```diff
@@ -0,0 +1,191 @@
+#  Copyright 2023 Red Hat, Inc.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import datetime
+import logging
+import os
+import pprint
+import re
+import threading
+import typing
+
+logger = logging.getLogger(__name__)
+
+
+###############################################################################
+class _Singleton(type):
+    def __init__(cls, name, bases, dct, **kwargs) -> None:
+        super().__init__(name, bases, dct, **kwargs)
+        cls.__singleton = None
+        cls.__lock = threading.RLock()
+
+    def __call__(cls, *args, **kwargs) -> type:
+        if not cls.__singleton:
+            with cls.__lock:
+                if not cls.__singleton:
+                    cls.__singleton = super().__call__(*args, **kwargs)
+        return cls.__singleton
+
+
+###############################################################################
+class Singleton(metaclass=_Singleton):
+    pass
+
+
+###############################################################################
+class DisplayBannerError(Exception):
+    pass
+
+
+###############################################################################
+class DisplayBannerIncompleteError(DisplayBannerError):
+    pass
+
+
+###############################################################################
+class Display(Singleton):
+    ###########################################################################
+    @classmethod
+    def instance(cls, level: int = None):
```

Comment:
If type is int, default should not be None.

Reply:
I mean the typehint is incorrect. If the type is int, the default value should be int. Otherwise it should be typing.optional[int]